### ISSUE TITLE
Breaking API check never fails

### ISSUE THREAD

**Issue by jakemac53** [dart-lang/ecosystem#304]
I am trying to add the breaking API check here https://github.com/dart-lang/build/pull/3760.

While in the comment it does notice the breaking change, and the job also sees the breaking change, it still passes, and also in the comment it is listed with a green checkmark.

I should be forced to do a breaking release?

**Comment by mosuem** [comment]
It seems like you found the fix already, adding `breaking` to the `fail_on` list in the settings.

**Comment by jakemac53** [comment]
> It seems like you found the fix already, adding `breaking` to the `fail_on` list in the settings.

I did this but it didn't actually  make the job fail still

**Comment by mosuem** [comment]
It does pass now, as the version is correct. Can you point to the invocation which incorrectly passed?

**Comment by jakemac53** [comment]
https://github.com/dart-lang/build/actions/runs/11150849769/job/30992945328?pr=3760 is the job which passes, but should be failing. It should be requiring a breaking release (5.0.0) but I only did a feature release.

**Comment by mosuem** [comment]
Why should it be failing? [The comment](https://github.com/dart-lang/build/actions/runs/11150849769/attempts/1#summary-30992945328) says it needs a minor version bump, which was made. What am I missing?

**Comment by jakemac53** [comment]
Oh interesting, it was definitely a breaking change.

I widened the return type of a function, so users of that function would be broken.

So, I guess the issue is actually that widening return types should be considered breaking and cause a major version.

**Comment by mosuem** [comment]
@devmil - Widening the return type seems to be a minor breaking change at the moment - should it be a major breaking change instead?

**Comment by devmil** [comment]
> @devmil - Widening the return type seems to be a minor breaking change at the moment - should it be a major breaking change instead?

I agree. I created an issue to look into this

**Comment by devmil** [comment]
> @devmil - Widening the return type seems to be a minor breaking change at the moment - should it be a major breaking change instead?

Widening the type already should lead to a breaking change and looking into the CI run (https://github.com/dart-lang/build/actions/runs/11150849769/job/30992945328?pr=3760) it looks like dart_apitool detected that change as breaking?
![CleanShot 2024-10-10 at 09 07 12@2x](https://github.com/user-attachments/assets/9f7cfcfc-48f2-4300-9650-f93811e31b01)

**Comment by mosuem** [comment]
Hm. But the `version` part of the report seems to indicate that the version should only be bumped from `4.1.0` to `4.2.0`, or am I overlooking something?

**Comment by mosuem** [comment]
Also, I thought `minor` indicates that this warrants only a minor version change, or what does it mean?

**Comment by devmil** [comment]
> Also, I thought `"type": "minor"` indicates that this warrants only a minor version change, or what does it mean?

Ah, I see. The confusion comes probably from the fact that the "type" in the JSON output is only relevant if breaking == "false" and tells if a minor version increase should happen. 
If the change is breaking then a major bump has to happen and the "type" field gets irrelevant.

I will change that so that type contains "major" in that situation which would make much more sense in my opinion

**Comment by mosuem** [comment]
Sounds good - and the `version` key of the report should then also suggest `5.0.0`, correct?

**Comment by devmil** [comment]
> Sounds good - and the `version` key of the report should then also suggest `5.0.0`, correct?

['version']['needed'] should already contain that information today

**Comment by mosuem** [comment]
Yes, the `needed` field is [the one being parsed](https://github.com/dart-lang/ecosystem/blob/bed43c312bc22d0c6f86c6d32e00235314c2a460/pkgs/firehose/lib/src/health/health.dart#L195).

**Comment by devmil** [comment]
> Yes, the `needed` field is [the one being parsed](https://github.com/dart-lang/ecosystem/blob/bed43c312bc22d0c6f86c6d32e00235314c2a460/pkgs/firehose/lib/src/health/health.dart#L195).

So, the problem is that the old version is a pre-release version. In that case there is no major bump requirement for breaking changes (dart_apitool only checks if the new version is equal or higher)

**Comment by mosuem** [comment]
Hm, that sounds like the API tool it is working correctly after all. 

The health PR check currently tests against the repository checkout, it's an interesting question if it should test against the latest released version to check if a major version bump is required instead. That would require downloading and unzipping the version from `pub.dev`.

@jakemac53 any opinion for your workflow in the build repo?

**Comment by devmil** [comment]
> Hm, that sounds like the API tool it is working correctly after all. 
> 
> The health PR check currently tests against the repository checkout, it's an interesting question if it should test against the latest released version to check if a major version bump is required instead. That would require downloading and unzipping the version from `pub.dev`.
> 
> @jakemac53 any opinion for your workflow in the build repo?

dart_apitool can do that already if the package is on pub.dev so you would just specify the "pub://package/version" Uri as the old package parameter
